edb/edgeql/compiler/triggers.py

```diff
@@ -0,0 +1,150 @@
+
+from edb.ir import ast as irast
+
+from edb.schema import name as sn
+from edb.schema import objtypes as s_objtypes
+from edb.schema import triggers as s_triggers
+
+from edb.edgeql import qltypes
+
+from . import context
+from . import dispatch
+from . import schemactx
+from . import setgen
+from . import typegen
+
+
+TRIGGER_KINDS = {
+    irast.UpdateStmt: qltypes.TriggerKind.Update,
+    irast.DeleteStmt: qltypes.TriggerKind.Delete,
+    irast.InsertStmt: qltypes.TriggerKind.Insert,
+}
+
+
+def compile_trigger(
+    trigger: s_triggers.Trigger,
+    affected: set[tuple[s_objtypes.ObjectType, irast.MutatingStmt]],
+    *,
+    ctx: context.ContextLevel,
+) -> irast.Trigger:
+    schema = ctx.env.schema
+
+    scope = trigger.get_scope(schema)
+    kinds = set(trigger.get_kinds(schema))
+    source = trigger.get_subject(schema)
+
+    with ctx.detached() as _, ctx.newscope(fenced=True) as sctx:
+        sctx.anchors = sctx.anchors.copy()
+
+        anchors = {}
+        new_path = irast.PathId.from_type(
+            schema, source, typename=sn.QualName(
+                module='__derived__', name='__new__')
+        )
+        new_set = setgen.class_set(
+            source, path_id=new_path, ignore_rewrites=True, ctx=sctx)
+
+        old_set = None
+        if qltypes.TriggerKind.Insert not in kinds:
+            old_path = irast.PathId.from_type(
+                schema, source, typename=sn.QualName(
+                    module='__derived__', name='__old__')
+            )
+            old_set = setgen.class_set(
+                source, path_id=old_path, ignore_rewrites=True, ctx=sctx)
+            old_set.expr = irast.TriggerAnchor(typeref=old_set.typeref)
+            anchors['__old__'] = old_set
+        if qltypes.TriggerKind.Delete not in kinds:
+            anchors['__new__'] = new_set
+
+        for name, ir in anchors.items():
+            if scope == qltypes.TriggerScope.Each:
+                sctx.path_scope.attach_path(ir.path_id, context=None)
+                sctx.iterator_path_ids |= {ir.path_id}
+            sctx.anchors[name] = ir
+
+        trigger_set = dispatch.compile(
+            trigger.get_expr(schema).qlast, ctx=sctx)
+
+    typeref = typegen.type_to_typeref(source, env=ctx.env)
+    taffected = {
+        (typegen.type_to_typeref(t, env=ctx.env), ir) for t, ir in affected
+    }
+
+    return irast.Trigger(
+        expr=trigger_set,
+        kinds=kinds,
+        scope=scope,
+        source_type=typeref,
+        affected=taffected,
+        new_set=new_set,
+        old_set=old_set,
+    )
+
+
+def compile_triggers(
+    dml_stmts: Collection[irast.MutatingStmt],
+    *,
+    ctx: context.ContextLevel,
+) -> tuple[irast.Trigger, ...]:
+    schema = ctx.env.schema
+
+    trigger_map: dict[
+        s_triggers.Trigger,
+        set[tuple[s_objtypes.ObjectType, irast.MutatingStmt]],
+    ] = {}
+    for stmt in dml_stmts:
+        kind = TRIGGER_KINDS[type(stmt)]
+
+        stype = schemactx.concretify(
```

Comment:
What does this do? I cannot gather from the function documentation. What are "views" in this context?

Reply:
"view" is what "shape" used to be called, and which it is still sometimes called internally. But more generally, all of the transient subtypes we generate during compilation.

`concretify` finds non-view concrete base types, but it does it recursively through unions and intersections.

Arguably this should be the normal thing? Idk.